### PR TITLE
Client-side app-less sandboxes

### PR DESCRIPTION
**Pull request by erikbern**
Implements the basics of this. Demo:

```
>>> from modal import Sandbox
>>> s = Sandbox.create("/bin/bash", "-c", "echo hello world")
>>> for line in s.stdout:
...     print(line)
... 
hello world

>>> s.poll()
0
```

TODO for this PR:
- [X] Support apps optionally
- [X] Add tests
- [X] Add tests for custom image and secrets

Will do in subsequent PRs:
* Update docstrings to use new syntax
* Add deprecation warnings for old syntax

### CODE REVIEW

File: modal/sandbox.py [hunk 1]
```diff
@@ -323,6 +330,63 @@ async def _load(self: _Sandbox, resolver: Resolver, _existing_object_id: Optiona
...
+            memory=memory,
+            network_file_systems=network_file_systems,
+            block_network=block_network,
+            volumes=volumes,
+            pty_info=pty_info,
+            _allow_background_volume_commits=_allow_background_volume_commits,
+            _experimental_scheduler_placement=_experimental_scheduler_placement,
+        )
+        app_id: Optional[str] = app.app_id if app else None
+        resolver = Resolver(client, environment_name=environment_name, app_id=app_id)
```
Comment (aksh-at) [inline modal/sandbox.py:386]: Are we able to resolve all deps without an app ID?

File: modal/sandbox.py [hunk 2]
```diff
@@ -323,6 +330,63 @@ async def _load(self: _Sandbox, resolver: Resolver, _existing_object_id: Optiona
...
+        # Specify, in MiB, a memory request which is the minimum memory required.
+        # Or, pass (request, limit) to additionally specify a hard limit in MiB.
+        memory: Optional[Union[int, Tuple[int, int]]] = None,
+        block_network: bool = False,  # Whether to block network access
+        volumes: Dict[
+            Union[str, os.PathLike], Union[_Volume, _CloudBucketMount]
+        ] = {},  # Mount points for Modal Volumes and CloudBucketMounts
+        pty_info: Optional[api_pb2.PTYInfo] = None,
+        _allow_background_volume_commits: Optional[bool] = None,
+        _experimental_scheduler_placement: Optional[
```
Comment (aksh-at) [inline modal/sandbox.py:357]: Ah looks like we still need to add the non-experimental scheduler param, now that it's out. Outside the scope of this PR.

File: modal/sandbox.py [hunk 3]
```diff
@@ -323,6 +330,63 @@ async def _load(self: _Sandbox, resolver: Resolver, _existing_object_id: Optiona
 
         return _Sandbox._from_loader(_load, "Sandbox()", deps=_deps)
 
+    @staticmethod
+    async def create(
+        *entrypoint_args: str,
```
Comment (aksh-at) [inline modal/sandbox.py:335]: Should we make the following args keyword-only?

**Review by aksh-at — APPROVED** [review]
LGTM! Should we add a deprecation message for the old code path?

File: modal/sandbox.py (hunk 3 above)
Comment (erikbern) [reply modal/sandbox.py:335]: yeah good point

File: modal/sandbox.py (hunk 1 above)
Comment (erikbern) [reply modal/sandbox.py:386]: yeah should be fine – it's just images that need it and i fixed that in the backend

Comment (erikbern) [reply modal/sandbox.py:386]: actually it's possible secrets need it too, let me check

File: modal/sandbox.py (hunk 3 above)
Comment (erikbern) [reply modal/sandbox.py:335]: actually believe they are keyword-only because of the * automatically (i just learned something new about python!)

Comment (aksh-at) [reply modal/sandbox.py:335]: oh cool, makes sense!

**Comment by erikbern** [comment]
Ok slightly bad news: in order for this to work, we need to support app-less secrets and mounts as well.

The good news is that this is something I've been hoping to add support for anyway for a while. But it will take a bit of time to figure out how to do it, unfortunately.

**Comment by erikbern** [comment]
> LGTM! Should we add a deprecation message for the old code path?

Was planning to do that in a subsequent PR (I want to update some integration tests and docs first)

**Comment by erikbern** [comment]
This _works_ now, but lack of output is a bit confusing with custom images.
```
>>> image = Image.debian_slim().pip_install("numpy").run_commands("sleep 30")
>>> s = Sandbox.create("/bin/bash", "-c", "echo hello world", image=image)  # This blocks for >30s
>>> s.poll()
0
>>> [line for line in s.stdout]
['hello world\n']
```

**Comment by aksh-at** [comment]
> This _works_ now, but lack of output is a bit confusing with custom images.

Ah yeah, I think we used to be able to see output in the app logs, but decoupling makes that difficult 🤔

**Comment by erikbern** [comment]
> Ah yeah, I think we used to be able to see output in the app logs, but decoupling makes that difficult 🤔

I think we can fix this in a different way. I've had it on my backlog for a long time to break out the output management into a singleton (separate from the resolver) that you can activate using a context manager (with the idea that all CLI commands enable this automatically). Something like

```python
with modal.enable_logs():
    ...
```

I don't love singletons but in this case I think it's fine since stdout is already a singleton

**Comment by erikbern** [comment]
ok this FINALLY works

```
>>> from modal import Image, Secret, Mount, Sandbox
>>> secret = Secret.from_dict({"FOO": "bar"})
>>> s = Sandbox.create("/bin/bash", "-c", "echo $FOO", secrets=[secret], environment_name="main")
>>> s.poll()
0
>>> [line for line in s.stdout]
['bar\n']
```

**Pull request merged**